### PR TITLE
Stack unwind fixes

### PR DESCRIPTION
**Pull request by bsdjhb**
Some of these are from the purecap branch.  kstack_contains() is new however, and I plan to upstream this entire series.  Alex ran into infinite recursion in the DDB stack unwinder for RISC-V that I think the last patch addresses.

I haven't yet tested this fully (still building)

### CODE REVIEW

**Comment by bsdjhb** [comment]
@arichardson did you have a way to test the bug you had earlier with a crash when unwinding the stack?

**Comment by arichardson** [comment]
> @arichardson did you have a way to test the bug you had earlier with a crash when unwinding the stack?

It happened while running the sys/netinet tests, I can't remember which one of them triggered the panic.

**Comment by arichardson** [comment]
I can no longer reproduce the panic with latest dev, but will try applying these patches on top of the commit that was entering the infinite backtrace loop.

**Review by arichardson — COMMENTED** [review]
I'm still seeing this inifinite panic loop with these changes cherry-picked on top of 31df7fe853dc87d7f23add1ae3fdf0194e7bf15c.


`cheribuild.py cheribsd-riscv64-purecap --skip-update --skip-buildworld --clean && cheribuild.py run-riscv64-purecap --test --test-extra-args="--no-run-cheribsdtest --kyua-tests-files=/usr/tests/sys/netinet/Kyuafile"`:
```
[PEXPECT\PROMPT]>kyua test --results-file=/tmp/results.db -k /usr/tests/sys/netinet/Kyuafile
carp:basic_v4  ->  skipped: This test requires carp  [2.353s]
carp:basic_v6  ->  skipped: This test requires carp  [2.398s]
divert:ipdivert_ip_input_local_success  ->  skipped: Required program 'scapy' not found in PATH  [0.079s]
divert:ipdivert_ip_output_remote_success  ->  skipped: Required program 'scapy' not found in PATH  [0.075s]
fibs_test:arpresolve_checks_interface_fib  ->  skipped: Required program 'nping' not found in PATH  [0.082s]
fibs_test:default_route_with_multiple_fibs_on_same_subnet  ->  skipped: This test requires net.add_addr_allfibs=0  [2.666s]
fibs_test:default_route_with_multiple_fibs_on_same_subnet_inet6  ->  skipped: This test requires net.add_addr_allfibs=0  [2.609s]
fibs_test:loopback_and_network_routes_on_nondefault_fib  ->  skipped: This test requires net.add_addr_allfibs=0  [2.624s]
fibs_test:loopback_and_network_routes_on_nondefault_fib_inet6  ->  skipped: This test requires net.add_addr_allfibs=0  [2.609s]
fibs_test:same_ip_multiple_ifaces  ->  skipped: The 0th configured fib is 1, which is not less than net.fibs, which is 1  [2.942s]
fibs_test:same_ip_multiple_ifaces_fib0  ->  passed  [7.018s]
fibs_test:same_ip_multiple_ifaces_inet6  ->  skipped: The 0th configured fib is 1, which is not less than net.fibs, which is 1  [2.927s]
fibs_test:slaac_on_nondefault_fib6  ->  skipped: This test requires net.add_addr_allfibs=0  [2.806s]
fibs_test:subnet_route_with_multiple_fibs_on_same_subnet  ->  skipped: This test requires net.add_addr_allfibs=0  [2.594s]
fibs_test:subnet_route_with_multiple_fibs_on_same_subnet_inet6  ->  skipped: This test requires net.add_addr_allfibs=0  [2.622s]
fibs_test:udp_dontroute  ->  skipped: This test requires net.add_addr_allfibs=0  [2.614s]
fibs_test:udp_dontroute6  ->  skipped: https://bugs.freebsd.org/244172  [2.651s]
forward:fwd_ip_icmp_gw_fast_success  ->  skipped: Required program 'scapy' not found in PATH  [0.076s]
forward:fwd_ip_icmp_gw_slow_success  ->  skipped: Required program 'scapy' not found in PATH  [0.077s]
forward:fwd_ip_icmp_iface_fast_success  ->  skipped: Required program 'scapy' not found in PATH  [0.076s]
forward:fwd_ip_icmp_iface_slow_success  ->  skipped: Required program 'scapy' not found in PATH  [0.076s]
ip_reass_test:ip_reass__large_fragment  ->  passed  [0.275s]
ip_reass_test:ip_reass__multiple_last_fragments  ->  passed  [0.231s]
ip_reass_test:ip_reass__zero_length_fragment  ->  passed  [0.218s]
lpm:lpm_test1_success  ->  epair initialized.
lock order reversal:
 1st 0xffffffc0007af520 allprison (allprison) @ /local/scratch/alr48/cheri/cheribsd/sys/kern/kern_jail.c:991
 2nd 0xffffffc0008810d0 vnet_sysinit_sxlock (vnet_sysinit_sxlock) @ /local/scratch/alr48/cheri/cheribsd/sys/net/vnet.c:577
stack backtrace:
#0 0xffffffc000314a74 at witness_checkorder+0xd6a
#1 0xffffffc0003147be at witness_checkorder+0xab4
#2 0xffffffc0002b9670 at _sx_slock_int+0x66
#3 0xffffffc0002b9abc at _sx_slock+0xc
#4 0xffffffc0003c81de at vnet_alloc+0xba
#5 0xffffffc000276920 at kern_jail_set+0x1692
#6 0xffffffc000277e20 at sys_jail_set+0x46
#7 0xffffffc0005ac330 at do_trap_user+0x1c4
#8 0xffffffc00059b6fe at cpu_exception_handler_user+0xde
#9 0xffffffc00059b6fe at cpu_exception_handler_user+0xde
#10 0xffffffc00059b6fe at cpu_exception_handler_user+0xde
#11 0xffffffc00059b6fe at cpu_exception_handler_user+0xde
#12 0xffffffc00059b6fe at cpu_exception_handler_user+0xde
#13 0xffffffc00059b6fe at cpu_exception_handler_user+0xde
#14 0xffffffc00059b6fe at cpu_exception_handler_user+0xde
#15 0xffffffc00059b6fe at cpu_exception_handler_user+0xde
#16 0xffffffc00059b6fe at cpu_exception_handler_user+0xde
#17 0xffffffc00059b6fe at cpu_exception_handler_user+0xde
ra = 0xffffffc0005aaea8
sp = 0xffffffc036d66220 [rwxRW,0x0000000000000000-0xffffffffffffffff]
gp = 0xffffffc036d66820 [rwxRW,0xa600000000000000-0x8c00000000000000] (invalid,sealed)
tp = 0x0000000000000000
t[0] = 0x00000000000003fd
t[1] = 0xffffffc036e3174c
t[2] = 0x0000000000000000
t[3] = 0xffffffc00028e8c8
t[4] = 0x0000000000000000
t[5] = 0x0000000000000000
t[6] = 0x0000000000000000
s[0] = 0xffffffc036d66230
s[1] = 0xffffffc0385f3900
s[2] = 0xffffffc00085ff08
s[3] = 0xffffffc036d66238
s[4] = 0xffffffc036d66250
s[5] = 0x0000000000000000
s[6] = 0x0000004000000000
s[7] = 0x0000000800000000
s[8] = 0xffffffffffffffff
s[9] = 0xffffffc00078c0d8
s[10] = 0xffffffc039200008
s[11] = 0x0000000000000fb0
a[0] = 0xffffffc036d66238
a[1] = 0xffffffc039200000
a[2] = 0x0000000000000001
a[3] = 0x0000000039200000
a[4] = 0x0000000800000000
a[5] = 0xffffffc00085fe60
a[6] = 0x0000000000002000
a[7] = 0x0000000000000001
sepc = 0xffffffc0005ad1c4 [rwxRW,0x0000000000000000-0xffffffffffffffff]
ddc = 0x0000000000000000 [rwxRW,0x0000000000000000-0xffffffffffffffff]
sstatus == 0x0000000000004100
stval == 0xffffffc0391ffff0
panic: Fatal page fault at 0xffffffc0005ad1c4: 0xffffffc0391ffff0
cpuid = 0
time = 1604577952

Trying to get a stack trace for kernel panic: <_sre.SRE_Match object; span=(5025, 5054), match='panic: Fatal page fault at 0x'>
KDB: stack backtrace:

Kernel stack trace about to be printed:
db_trace_self() at db_trace_self
db_fetch_ksymtab() at db_fetch_ksymtab+0x146
kdb_backtrace() at kdb_backtrace+0x2c
vpanic() at vpanic+0x146
panic() at panic+0x52
do_trap_supervisor() at do_trap_supervisor+0x5c8
do_trap_supervisor() at do_trap_supervisor+0x94
cpu_exception_handler_supervisor() at cpu_exception_handler_supervisor+0xe2
--- exception 13, tval = 0xffffffc0391ffff0
unwind_frame() at unwind_frame+0x1c
stack_save() at stack_save+0x7c
witness_checkorder() at witness_checkorder+0x45c
__mtx_lock_flags() at __mtx_lock_flags+0x9a
.LBB6_59() at .LBB6_59+0x24
ra = 0xffffffc00059ae32
sp = 0xffffffc036d65a80 [rwxRW,0x0000000000000000-0xffffffffffffffff]
gp = 0xffffffc036d65a20 [rwxRW,0xffffffbffc400000-0xffffffc08c000000] (invalid,sealed)
tp = 0xffffffc036d658c0 [rwxRW,0xff00000000000000-0x0400000000000000] (invalid,sealed)
t[0] = 0x0000000000000000
t[1] = 0x0000000000000000
t[2] = 0x0000000000000000
t[3] = 0x0000000000000000
t[4] = 0x0000000000000000
t[5] = 0x0000000000000000
t[6] = 0x0000000000000000
s[0] = 0xffffffc036d65a90
s[1] = 0xffffffc036d65ff0
s[2] = 0xffffffc0385f3900
s[3] = 0xffffffc000617755
s[4] = 0xffffffc000613f8e
s[5] = 0xffffffc0005f9ad9
s[6] = 0x7fffffffffffffff
s[7] = 0x0000000800000001
s[8] = 0xffffffc00060fee6
s[9] = 0x0000004000000000
s[10] = 0xffffffc036d65b38
s[11] = 0xffffffc00061cdf4
a[0] = 0xffffffc036d65b38
a[1] = 0xffffffc039200000
a[2] = 0x0000000000000001
a[3] = 0x0000000039200000
a[4] = 0x0000000800000000
a[5] = 0xffffffc0005970da
a[6] = 0x0000000000000000
a[7] = 0xffffffc00078c0e8
sepc = 0xffffffc0005ad1c4 [rwxRW,0x0000000000000000-0xffffffffffffffff]
ddc = 0x0000000000000000 [rwxRW,0x0000000000000000-0xffffffffffffffff]
sstatus == 0x0000000000004100
stval == 0xffffffc0391ffff0
panic: Fatal page fault at 0xffffffc0005ad1c4: 0xffffffc0391ffff0
cpuid = 0
time = 1604577952
KDB: stack backtrace:
db_trace_self() at db_trace_self
db_fetch_ksymtab() at db_fetch_ksymtab+0x146
kdb_backtrace() at kdb_backtrace+0x2c
vpanic() at vpanic+0x146
panic() at panic+0x52
do_trap_supervisor() at do_trap_supervisor+0x5c8
do_trap_supervisor() at do_trap_supervisor+0x94
cpu_exception_handler_supervisor() at cpu_exception_handler_supervisor+0xe2
--- exception 13, tval = 0xffffffc0391ffff0
unwind_frame() at unwind_frame+0x1c
db_trace_thread() at db_trace_thread+0x19e
db_trace_self() at db_trace_self+0x2a
db_fetch_ksymtab() at db_fetch_ksymtab+0x146
kdb_backtrace() at kdb_backtrace+0x2c
vpanic() at vpanic+0x146
panic() at panic+0x52
do_trap_supervisor() at do_trap_supervisor+0x5c8
do_trap_supervisor() at do_trap_supervisor+0x94
cpu_exception_handler_supervisor() at cpu_exception_handler_supervisor+0xe2
--- exception 13, tval = 0xffffffc0391ffff0
unwind_frame() at unwind_frame+0x1c
stack_save() at stack_save+0x7c
witness_checkorder() at witness_checkorder+0x45c
__mtx_lock_flags() at __mtx_lock_flags+0x9a
.LBB6_59() at .LBB6_59+0x24
ra = 0xffffffc00059ae32
sp = 0xffffffc036d652e0 [rwxRW,0x0000000000000000-0xffffffffffffffff]
gp = 0xffffffc036d65280 [rwxRW,0xffffffbffc400000-0xffffffc08c000000] (invalid,sealed)
tp = 0xffffffc036d65120 [rwxRW,0xff00000000000000-0x0400000000000000] (invalid,sealed)
t[0] = 0x0000000000000000
t[1] = 0x0000000000000000
t[2] = 0x0000000000000000
t[3] = 0x0000000000000000
t[4] = 0x0000000000000000
t[5] = 0x0000000000000000
t[6] = 0x0000000000000000
s[0] = 0xffffffc036d652f0
s[1] = 0xffffffc036d65ff0
s[2] = 0xffffffc0385f3900
s[3] = 0xffffffc000617755
s[4] = 0xffffffc000613f8e
s[5] = 0xffffffc0005f9ad9
s[6] = 0x7fffffffffffffff
s[7] = 0x0000000800000001
s[8] = 0xffffffc00060fee6
s[9] = 0x0000004000000000
s[10] = 0xffffffc036d65398
s[11] = 0xffffffc00061cdf4
a[0] = 0xffffffc036d65398
a[1] = 0xffffffc039200000
a[2] = 0x0000000000000001
a[3] = 0x0000000039200000
a[4] = 0x0000000800000000
a[5] = 0xffffffc0005970da
a[6] = 0x0000000000000000
a[7] = 0xffffffc00078c0e8
sepc = 0xffffffc0005ad1c4 [rwxRW,0x0000000000000000-0xffffffffffffffff]
ddc = 0x0000000000000000 [rwxRW,0x0000000000000000-0xffffffffffffffff]
sstatus == 0x0000000000004100
stval == 0xffffffc0391ffff0
panic: Fatal page fault at 0xffffffc0005ad1c4: 0xffffffc0391ffff0
cpuid = 0
time = 1604577952
KDB: stack backtrace:

Unexpected output (infinite backtrace loop?): KDB: stack backtrace:

GOT KERNEL PANIC!

EXITING DUE TO KERNEL PANIC!
Running rm -fv /local/scratch/alr48/cheri/output/cheribsd-riscv64-purecap.img.runtests.20201105120053.pid34804
removed '/local/scratch/alr48/cheri/output/cheribsd-riscv64-purecap.img.runtests.20201105120053.pid34804'
```

File: sys/mips/mips/stack_machdep.c
Comment (arichardson) [inline sys/mips/mips/stack_machdep.c:169]: will __builtin_frame_address() also work, or does this need to be the current SP? 
__builtin_frame_address() will do the right thing for purecap too.

Comment (jrtc27) [reply sys/mips/mips/stack_machdep.c:169]: PowerPC uses just `__builtin_frame_address(0)`, whilst Arm, Arm64 and RISC-V use both. X86 only uses the frame pointer, but obtained through inline assembly. It's a mess upstream.

Comment (bsdjhb) [reply sys/mips/mips/stack_machdep.c:169]: In the purecap kernel there was already a separate stack_machdep.c and I changed it to use `c11`.  It may be that builtin_frame_address would work.  I haven't tried it.

**Comment by bsdjhb** [comment]
I actually merged some of these changes upstream already, so I will need to wait to rebase this after we merge in August 12 commits from upstream.  I will still push an updated version of this in a bit that I think fixes the latest panic you saw.

**Comment by bsdjhb** [comment]
@arichardson this should be testable now, but I will wait to merge it until I can rebase it on top of some upstream changes that we need to loop back into dev.

File: sys/sys/proc.h
Comment (arichardson) [inline sys/sys/proc.h:1210]: I feel like surely there must already be a helper function for "range is subset of another range" somewhere, but a quick search doesn't show me anything.

File: sys/mips/mips/stack_machdep.c
Comment (arichardson) [inline sys/mips/mips/stack_machdep.c:49]:
```suggestion

	return (kstack_contains(td, sp + stack_pos, sizeof(u_register_t)));
```

Maybe inline this variable?

**Review by arichardson — APPROVED** [review]
🎉 I can confirm that this fixes the inifite panic loop

```
kyua test --results-file=/tmp/results.db -k /usr/tests/sys/netinet/Kyuafile
carp:basic_v4  ->  skipped: This test requires carp  [2.399s]
carp:basic_v6  ->  skipped: This test requires carp  [2.445s]
divert:ipdivert_ip_input_local_success  ->  skipped: Required program 'scapy' not found in PATH  [0.081s]
divert:ipdivert_ip_output_remote_success  ->  skipped: Required program 'scapy' not found in PATH  [0.076s]
fibs_test:arpresolve_checks_interface_fib  ->  skipped: Required program 'nping' not found in PATH  [0.093s]
fibs_test:default_route_with_multiple_fibs_on_same_subnet  ->  skipped: This test requires net.add_addr_allfibs=0  [2.775s]
fibs_test:default_route_with_multiple_fibs_on_same_subnet_inet6  ->  skipped: This test requires net.add_addr_allfibs=0  [2.754s]
fibs_test:loopback_and_network_routes_on_nondefault_fib  ->  skipped: This test requires net.add_addr_allfibs=0  [2.701s]
fibs_test:loopback_and_network_routes_on_nondefault_fib_inet6  ->  skipped: This test requires net.add_addr_allfibs=0  [2.810s]
fibs_test:same_ip_multiple_ifaces  ->  skipped: The 0th configured fib is 1, which is not less than net.fibs, which is 1  [3.061s]
fibs_test:same_ip_multiple_ifaces_fib0  ->  passed  [7.474s]
fibs_test:same_ip_multiple_ifaces_inet6  ->  skipped: The 0th configured fib is 1, which is not less than net.fibs, which is 1  [3.761s]
fibs_test:slaac_on_nondefault_fib6  ->  skipped: This test requires net.add_addr_allfibs=0  [2.994s]
fibs_test:subnet_route_with_multiple_fibs_on_same_subnet  ->  skipped: This test requires net.add_addr_allfibs=0  [2.737s]
fibs_test:subnet_route_with_multiple_fibs_on_same_subnet_inet6  ->  skipped: This test requires net.add_addr_allfibs=0  [2.725s]
fibs_test:udp_dontroute  ->  skipped: This test requires net.add_addr_allfibs=0  [2.706s]
fibs_test:udp_dontroute6  ->  skipped: https://bugs.freebsd.org/244172  [2.746s]
forward:fwd_ip_icmp_gw_fast_success  ->  skipped: Required program 'scapy' not found in PATH  [0.067s]
forward:fwd_ip_icmp_gw_slow_success  ->  skipped: Required program 'scapy' not found in PATH  [0.079s]
forward:fwd_ip_icmp_iface_fast_success  ->  skipped: Required program 'scapy' not found in PATH  [0.066s]
forward:fwd_ip_icmp_iface_slow_success  ->  skipped: Required program 'scapy' not found in PATH  [0.076s]
ip_reass_test:ip_reass__large_fragment  ->  passed  [0.205s]
ip_reass_test:ip_reass__multiple_last_fragments  ->  passed  [0.251s]
ip_reass_test:ip_reass__zero_length_fragment  ->  passed  [0.204s]
lpm:lpm_test1_success  ->  epair initialized.
lock order reversal:
 1st 0xffffffc0007af520 allprison (allprison) @ /local/scratch/alr48/cheri/cheribsd/sys/kern/kern_jail.c:991
 2nd 0xffffffc0008810d0 vnet_sysinit_sxlock (vnet_sysinit_sxlock) @ /local/scratch/alr48/cheri/cheribsd/sys/net/vnet.c:577
stack backtrace:
#0 0xffffffc000314a74 at witness_checkorder+0xd6a
#1 0xffffffc0003147be at witness_checkorder+0xab4
#2 0xffffffc0002b9670 at _sx_slock_int+0x66
#3 0xffffffc0002b9abc at _sx_slock+0xc
#4 0xffffffc0003c81de at vnet_alloc+0xba
#5 0xffffffc000276920 at kern_jail_set+0x1692
#6 0xffffffc000277e20 at sys_jail_set+0x46
#7 0xffffffc0005ac2c0 at do_trap_user+0x1c4
#8 0xffffffc00059b6ee at cpu_exception_handler_user+0xde
passed  [21.681s]
lpm:lpm_test2_success  ->  passed  [23.306s]
```

**Comment by arichardson** [comment]
Plain RISC-V is broken though:

```
/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-727/cheribsd/sys/cddl/dev/dtrace/riscv/dtrace_isa.c:93:26: error: too few arguments to function call, expected 2, have 1
                if (unwind_frame(&state))
                    ~~~~~~~~~~~~       ^
./machine/stack.h:49:6: note: 'unwind_frame' declared here
bool unwind_frame(struct thread *, struct unwind_state *);
     ^
In file included from /local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-727/cheribsd/sys/cddl/contrib/opensolaris/uts/common/dtrace/dtrace.c:18451:
/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-727/cheribsd/sys/cddl/dev/dtrace/riscv/dtrace_isa.c:274:29: error: too few arguments to function call, expected 2, have 1
                done = unwind_frame(&state);
                       ~~~~~~~~~~~~       ^
./machine/stack.h:49:6: note: 'unwind_frame' declared here
bool unwind_frame(struct thread *, struct unwind_state *);
     ^
2 errors generated.
```

**Comment by bsdjhb** [comment]
I'll fix the dtrace breakage after I rebase it.

**Comment by bsdjhb** [comment]
dtrace breakage should be fixed, and I've also mirrored the changes to aarch64.  Still need to wait to rebase this on top of some upstream changes from Aug 21 before this is a merge candidate though.

File: sys/mips/mips/stack_machdep.c
Comment (bsdjhb) [reply sys/mips/mips/stack_machdep.c:49]: Maybe I should just inline this entire function.  Originally I didn't have kstack_contains so this function was slightly larger, but maybe it's redundant now.

File: sys/sys/proc.h
Comment (bsdjhb) [reply sys/sys/proc.h:1210]: No, it's open-coded in various places though. :-/

**Comment by arichardson** [comment]
```
>>> stage 3.1: building everything
--------------------------------------------------------------
/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-727/cheribsd/sys/ddb/db_ps.c:530:38: error: expected ')'
                        if (kstack_contains(td, saddr, 1) {
                                                          ^
/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-727/cheribsd/sys/ddb/db_ps.c:530:7: note: to match this '('
                        if (kstack_contains(td, saddr, 1) {
                           ^
/local/scratch/jenkins/workspace/CheriBSD-pipeline_PR-727/cheribsd/sys/ddb/db_ps.c:534:3: error: expected statement
                }
                ^
```

File: sys/mips/mips/stack_machdep.c
Comment (bsdjhb) [reply sys/mips/mips/stack_machdep.c]: I inlined the function.

**Comment by bsdjhb** [comment]
This is ready to merge (or if anyone has any additional review).  I will also be pushing most of this to FreeBSD as well.

File: sys/cddl/dev/dtrace/mips/dtrace_isa.c
Comment (arichardson) [inline sys/cddl/dev/dtrace/mips/dtrace_isa.c:219]: Not 100% sure this will always do the right thing for purecap. It should do for MIPS and RISC-V but not sure about Morello.
I'm not sure whether address-of-label needs to give a valid pointer or if an offset relative to entry point is also fine. I think as long as `goto *foo;` works anything is fine.

**Review by arichardson — APPROVED** [review]

File: sys/cddl/dev/dtrace/mips/dtrace_isa.c
Comment (bsdjhb) [reply sys/cddl/dev/dtrace/mips/dtrace_isa.c:219]: I do think this will need some refinement for purecap, if nothing else we will have to unseal it for at least MIPS since we read the instruction to figure out the stack layout.  The current MIPS purecap kernel has its own stack_machdep.c and doesn't support DTrace or DDB, so it hasn't really had to fully wrestle with this problem yet.  It does though already have to deal with unsealing return addresses.

Do note that the use of address-of-label is MIPS only.  RISC-V and aarch64 use different approaches.

**Pull request merged**
